Subclasses of Twisted's protocol fail to be determined by mypy-zope

Thanks for making mypy-zope! We've been using it for Synapse (which uses Twisted) and ran into a hiccup.

The following will cause mypy-zope to emit an error:

> error: Cannot determine consistent method resolution order (MRO) for "Foo"  [misc]

```python
from twisted.internet import protocol

class Foo(protocol.Protocol):
    pass

reveal_type(Foo)
```

The class hierarchy looks like `Foo` -> `Protocol` (which implements `IProtocol` and `ILoggingContext`) -> `BaseProtocol`.

I tried to create a reduced test-case for this, but was not able to, so it might be specific to something Twisted has done. Any help would be appreciated!

> Do you get this error when running mypy with cold caches (i.e. first time after you delete `.mypy_cache` directory?

Nope!

```sh
$ rm -r .mypy_cache
$ mypy temp.py
temp.py:6: note: Revealed type is 'def () -> temp.Foo'
$ mypy temp.py
temp.py:3: error: Cannot determine consistent method resolution order (MRO) for "Foo"  [misc]
temp.py:6: note: Revealed type is 'def () -> temp.Foo'
Found 1 error in 1 file (checked 1 source file)
```

> Do you get this error if you are running mypy daemon? (i.e. by running `dmypy run`)

I don't think so, but I've never used this before so I'm not 100% sure I'm using it properly.

--- a/src/mypy_zope/plugin.py
+++ b/src/mypy_zope/plugin.py
@@ -98,6 +98,8 @@ class ZopeInterfacePlugin(Plugin):
     def __init__(self, options: Options):
         super().__init__(options)
         self.fallback = DefaultPlugin(options)
+        # A cache of interface fullname to the fake implementation.
+        self._interfaces: Dict[str, TypeInfo] = {}
 
     def log(self, msg: str) -> None:
         if self.options.verbosity >= 1:
@@ -249,11 +251,16 @@ class ZopeInterfacePlugin(Plugin):
                 f"{iface_type.fullname}: {class_info.fullname}"
             )
 
-            # Make sure implementation is treates subtype of an interface. Pretend
-            # there is a decorator for the class that will create a "type promotion"
-            faketi = TypeInfo(SymbolTable(), iface_type.defn, iface_type.module_name)
-            faketi._promote = Instance(iface_type, [])
-            class_info.mro.append(faketi)
+            # Make sure implementation is treated as a subtype of an interface. Pretend
+            # there is a decorator for the class that will create a "type promotion",
+            # but ensure this only gets applied a single time per interface.
+            faketi = self._interfaces.get(iface_type.fullname)
+            if not faketi:
+                faketi = TypeInfo(SymbolTable(), iface_type.defn, iface_type.module_name)
+                faketi._promote = Instance(iface_type, [])
+                self._interfaces[iface_type.fullname] = faketi
+            if faketi not in class_info.mro:
+                class_info.mro.append(faketi)
 
         def analyze(classdef_ctx: ClassDefContext) -> None:
             api = classdef_ctx.api

--- a/src/mypy_zope/plugin.py
+++ b/src/mypy_zope/plugin.py
@@ -98,8 +98,6 @@ class ZopeInterfacePlugin(Plugin):
     def __init__(self, options: Options):
         super().__init__(options)
         self.fallback = DefaultPlugin(options)
-        # A cache of interface fullname to the fake implementation.
-        self._interfaces: Dict[str, TypeInfo] = {}
 
     def log(self, msg: str) -> None:
         if self.options.verbosity >= 1:
@@ -254,12 +252,10 @@ class ZopeInterfacePlugin(Plugin):
             # Make sure implementation is treated as a subtype of an interface. Pretend
             # there is a decorator for the class that will create a "type promotion",
             # but ensure this only gets applied a single time per interface.
-            faketi = self._interfaces.get(iface_type.fullname)
-            if not faketi:
+            promote = Instance(iface_type, [])
+            if not any(ti._promote == promote for ti in class_info.mro):
                 faketi = TypeInfo(SymbolTable(), iface_type.defn, iface_type.module_name)
-                faketi._promote = Instance(iface_type, [])
-                self._interfaces[iface_type.fullname] = faketi
-            if faketi not in class_info.mro:
+                faketi._promote = promote
                 class_info.mro.append(faketi)
 
         def analyze(classdef_ctx: ClassDefContext) -> None:

